llms/mlx_lm/utils.py

```diff
@@ -143,17 +143,21 @@ def sample(logits: mx.array) -> Tuple[mx.array, float]:
         if temp == 0:
             token = mx.argmax(logits, axis=-1)
         else:
+            # referenced implementation from https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L449-L460
             if top_p > 0 and top_p < 1.0:
                 if (
                     logits.dtype == mx.bfloat16
                 ):  # workdaround for unable to load kernel contiguous_scan_inclusive_sum_bfloat16_bfloat16
                     logits = logits.astype(mx.float32)
+
```

Comment:
Do you mind refactoring this `top_p` into a little function we can add a test for it?

Reply:
Sure thing, could you help double-check if the implementation makes sense to you now? Once we are confident in the implementation, I can add tests for it.